qiskit_experiments/database_service/db_experiment_data.py

```diff
@@ -930,80 +1018,99 @@ def block_for_results(self, timeout: Optional[float] = None) -> "DbExperimentDat
-            LOG.warning("Possibly incomplete experiment data: A Job returned an error.")
-
-    def _wait_for_callbacks(self, timeout: Optional[float] = None):
+        if excepts:
+            LOG.error("Job futures raised exceptions [eid: %s]:%s", self.experiment_id, excepts)
+
+    def _wait_for_analysis(self, timeout: Optional[float] = None):
         """Wait for analysis callbacks to finish"""
-        # Wait for analysis callbacks to finish
-        if self._callback_statuses:
-            for status in self._callback_statuses.values():
-                if status.status in [JobStatus.DONE, JobStatus.CANCELLED]:
-                    continue
-                LOG.info("Waiting for analysis callback %s to finish.", status.callback)
-                finished, timeout = combined_timeout(status.event.wait, timeout)
-                if not finished:
-                    LOG.warning(
-                        "Possibly incomplete analysis results:"
-                        " analysis"
-                        " callback %s timed out.",
-                        status.callback,
-                    )
+        waited = futures.wait(self._analysis_futures.values(), timeout=timeout)
+        # Log jobs still running after timeout
+        if waited.not_done:
+            LOG.info(
+                "Waiting for analysis callbacks timed out before all "
+                "callbacks completed [eid: %s].",
+                self.experiment_id,
+            )
 
-        # Check analysis status and show warning if cancelled or error
-        callback_status = self._callback_status()
-        if callback_status == "CANCELLED":
-            LOG.warning("Possibly incomplete analysis results: an analysis callback was cancelled.")
-        elif callback_status == "ERROR":
-            LOG.warning(
-                "Possibly incomplete analysis results: an analysis callback raised an error."
+        # Check for jobs that were cancelled or errorred
+        excepts = ""
+        for fut in waited.done:
+            ex = fut.exception()
+            if ex:
+                excepts += "\n".join(traceback.format_exception(type(ex), ex, ex.__traceback__))
+            elif fut.cancelled():
+                LOG.debug(
+                    "Analysis callbacks were cancelled before completion [eid: %s]",
+                    self.experiment_id,
+                )
+        if excepts:
+            LOG.error(
+                "Analysis callbacks raised exceptions [eid: %s]:%s", self.experiment_id, excepts
             )
 
-    def status(self) -> str:
-        """Return the data processing status.
+    def _removed_done_futures(self):
+        """Remove futures that have finished"""
+
+        def _keep_future(fut):
+            return not fut.done() or fut.cancelled() or fut.exception()
 
-        If the experiment consists of multiple jobs, the returned status is mapped
-        in the following order:
+        with self._analysis_futures.lock and self._job_futures.lock:
```

Comment:
We don't need to shutdown executor when `not_done_*` is empty?

Reply:
The executor doesn't need to be shutdown, it can sit around forever waiting to make more futures if needed.